File: 044.rb
# 044
# Wildcard Matching
# 2015-06-23
#####################################################
# @param {String} s
# @param {String} p
# @return {Boolean}
def is_match(s, p)
  l1 = s.size
  return false if l1 < p.size - p.count('*')

  m = Array.new(l1+1) {false}
  m[0] = true

  p.each_char do |c|
    if c == '*'
      1.upto(l1) do |i|
        m[i] = (m[i-1] or m[i])
      end
    else
      l1.downto(1) do |i|
        m[i] = (m[i-1] and (s[i-1] == c or c == '?'))
      end
      m[0] = false
    end
  end
  m[l1]
end

#####################################################
s = "aa"
p = "a*"

puts s
puts p
puts is_match(s, p)

File: 044.py
#   044
#   Wildcard Matching
#   2015-06-20
#####################################################
class Solution:
    # @param {string} s
    # @param {string} p
    # @return {boolean}
    def isMatch(self, s, p):
        l1 = len(s)
        l2 = len(p)
        if l1 < l2 - p.count('*'):
            return False

        m = [True] + [False] * l1
        for j in range(0, l2):
            if p[j] == '*':
                for i in range(1, l1+1):
                    m[i] = m[i-1] or m[i]
            else:
                for i in range(l1, 0, -1):
                    m[i] = m[i-1] and (s[i-1] == p[j] or p[j] == '?')
                m[0] = False

        return m[l1]

#####################################################

s = "abc"
p = "*ab*"


sol = Solution()
print sol.isMatch(s, p)   

